packages/typescript-estree/src/parseSettings/index.ts

```diff
@@ -57,6 +58,13 @@ export interface MutableParseSettings {
    */
   errorOnUnknownASTType: boolean;
 
+  /**
+   * Experimental: TypeScript server to power program creation.
+   */
+  EXPERIMENTAL_projectService:
+    | tsserverlibrary.server.ProjectService
```

Comment:
Will including this have a negative effect for those who are importing this, since now two big TS d.ts files will get loaded? Maybe not too many use this, but, I'm just thinking ahead.

Reply:
Hmm, I hope not! So far it hasn't hurt anything I've noticed.

Reply:
This will actually leak out into the public types accidentally.
`index` has `export { createProgramFromConfigFile as createProgram } from './create-program/useProvidedPrograms';`, and `./create-program/useProvidedPrograms` has `declare function useProvidedPrograms(programInstances: Iterable<ts.Program>, parseSettings: ParseSettings): ASTAndProgram | undefined;` - i.e. `import type { ParseSettings } from '../parseSettings'`

To properly isolate this Josh we'll probably want to refactor `create-program/useProvidedPrograms` into two files so that we don't import it from the `index` - i.e. just have `createProgramFromConfigFile` in its own file so that we don't leak the internal types.

I was poking around via `unpkg` but might be worth just double checking all of the paths with intellisense from a local build to double check that is the only place it leaks into `index`'s dependency graph.

Reply:
I see what you mean - as a quicker / less change-y step, I just changed the parameters for these functions to be their own types (or just a `string`). How does that feel to you?

Reply:
that fixes the issue, yeah because we no longer leak `ParseSettings` into a file that's imported via the `index` - meaning consumers won't ever implicitly load the tssserverlibrary defs.

Reply:
This file still contains the import from `tsserverlibrary`; from the above it seemed like it was able to be eliminated. Is that still the case?

Reply:
We're just using it here for `EXPERIMENTAL_projectService: tsserverlibrary.server.ProjectService | undefined;`, and it's a `type` import. I'd think this is enough to fix the issues around importing both TypeScript files? Unless we're talking about different things?

Reply:
I was actually originally referring to the type checking, in that uses of your API will type check TS's huge API twice.

Reply:
Is there any alternative that would prevent this from being visible externally, or is this definitely a requirement?

Reply:
🤔 long-term I think we really want this to be visible externally. I want to investigate integrating with other project service consumers such as https://github.com/Quramy/typescript-eslint-language-service (cc @Quramy).

I don't think I follow what you mean by users of our API type checking the huge API twice - do you mean because if this file's `.d.ts` output is in the `node_modules/` of projects, TypeScript (e.g. via `tsc`) will type check it?

If so: I suppose in theory we could hand-write a `ProjectService` type that coincidentally matches the `tsserverlibrary.server.ProjectService` type, with comments explaining why. Does that feel reasonable to you?

Reply:
If someone imports `@typescript-eslint/estree`, they will then access the `d.ts` files for this, which then references both `typescript` and `tsserverlibrary`, and since they're not the same file and both large, it'll increase the check time of anyone who imports this package.

Probably I should revisit the "merge typescript and tsserverlibrary" idea, I had just been working on other stuff.

Reply:
Ah, got it thanks. I think this is fine for now. I'm under the impression that very few consumers import from any `@typescript-eslint/*` package - and it's mostly authors of custom rules. Good to note as a potential followup if someone reports anything.